libs/sdk-core/src/lib/db/search/agentic.ts

```diff
@@ -0,0 +1,439 @@
+
+  export interface Step {
+    action: Action;
+    next?: { if?: string; stepId: string }[];
+    statusMessage?: string;
+  }
+
+  export interface Event {
+    stepId: string;
+    status: 'running' | 'done' | 'error';
+    message?: string;
+  }
+
+  export type Steps = { [stepId: string]: Step };
+
+  export class Pipeline {
+    kb: KnowledgeBox;
+    steps: { [stepId: string]: Step };
+    context: { [id: string]: any } = {};
+    private _events = new Subject<{ stepId: string; params: UserEventAction }>();
+    events = this._events.asObservable();
+    userResponses = new Subject<{ stepId: string; value: boolean | string | string[] }>();
+    private _status = new Subject<Event>();
+    status = this._status.asObservable();
+
+    constructor(
+      private _kb: KnowledgeBox,
+      private _steps: Steps,
+    ) {
+      this.kb = _kb;
+      this.steps = _steps;
+    }
+
+    run(initialContext: object): Observable<boolean> {
+      this.context = initialContext;
+      return this.runStep('START');
+    }
+
+    private runStep(stepId: string): Observable<boolean> {
+      const step = this.steps[stepId];
+      if (!step) {
+        return of(false);
+      }
+      this._status.next({ stepId, status: 'running' });
+      return this.runSingleStep(stepId).pipe(
+        switchMap((success) => {
+          if (success) {
+            const message = step.statusMessage ? this.format(step.statusMessage) : '';
+            this._status.next({ stepId, status: 'done', message });
+            const nextSteps = (step.next || [])
+              .filter((step) => !step.if || this.eval<boolean>(step.if))
+              .map((next) => next.stepId);
+
+            return nextSteps.length === 0
+              ? of(true)
+              : forkJoin(nextSteps.map((nextStepId) => this.runStep(nextStepId).pipe(take(1)))).pipe(
+                  map((results) => results.every((r) => r)),
+                );
+          } else {
+            this._status.next({ stepId, status: 'error' });
+            return of(false);
+          }
+        }),
+      );
+    }
+
+    private runSingleStep(stepId: string): Observable<boolean> {
+      const step = this.steps[stepId];
+      if (!step) {
+        return of(false);
+      }
+      if (step.action.type === 'predict') {
+        return this.runPredictAction(stepId, step.action);
+      } else if (step.action.type === 'find') {
+        return this.runFindAction(stepId, step.action);
+      } else if (step.action.type === 'ask') {
+        return this.runAskAction(stepId, step.action);
+      } else if (step.action.type === 'web') {
+        return this.runWebAction(stepId, step.action);
+      } else if (step.action.type === 'api') {
+        return this.runAPIAction(stepId, step.action);
+      } else if (step.action.type === 'user') {
+        return this.runUserAction(stepId, step.action);
+      }
+      return of(true);
```

Comment:
a switch would be either to read and maintain here (typescript is raising an error when a switch doesn't contain all the possible options)